drivers/timer/cavs_timer.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <arch/xtensa/xtensa_rtos.h>
+
+/**
+ * @file
+ * @brief CAVS DSP Wall Clock Timer driver
+ *
+ * The CAVS DSP on Intel SoC has a timer with one counter and two compare
+ * registers that is external to the CPUs. This timer is accessible from
+ * all available CPU cores and provides a synchronized timer under SMP.
+ */
+
+#define TIMER		0
+#define TIMER_IRQ	DSP_WCT_IRQ(TIMER)
+#define CYC_PER_TICK	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC	\
+			/ CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define MAX_CYC		0xFFFFFFFFUL
+#define MAX_TICKS	((MAX_CYC - CYC_PER_TICK) / CYC_PER_TICK)
+#define MIN_DELAY	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 100000)
+
+static struct k_spinlock lock;
+static u64_t last_count;
+
+static volatile struct soc_dsp_shim_regs *shim_regs =
+	(volatile struct soc_dsp_shim_regs *)SOC_DSP_SHIM_REG_BASE;
+
+static void set_compare(u64_t time)
+{
+#if (TIMER == 0)
+	/* Set compare register */
+	shim_regs->dspwct0c = time;
+#elif (TIMER == 1)
+	/* Set compare register */
+	shim_regs->dspwct1c = time;
+#else
+#error "TIMER has to be 0 or 1!"
+#endif
+
+	/* Arm the timer */
+	shim_regs->dspwctcs |= DSP_WCT_CS_TA(TIMER);
+}
+
+static u64_t count(void)
+{
+	return shim_regs->walclk;
+}
+
+static u32_t count32(void)
+{
+	return shim_regs->walclk32_lo;
+}
+
+static void compare_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+	u64_t curr;
+	u32_t dticks;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	curr = count();
+
+#ifdef CONFIG_SMP
+	/* If it has been too long since last_count,
+	 * this interrupt is likely the same interrupt
+	 * event but being processed by another CPU.
+	 * Since it has already been processed and
+	 * ticks announced, skip it.
+	 */
+	if ((count32() - (u32_t)last_count) < MIN_DELAY) {
+		k_spin_unlock(&lock, key);
+		return;
+	}
+#endif
+
+	dticks = (u32_t)((curr - last_count) / CYC_PER_TICK);
+
+	/* Clear the triggered bit */
+	shim_regs->dspwctcs |= DSP_WCT_CS_TT(TIMER);
+
+	last_count += dticks * CYC_PER_TICK;
+
+#ifndef CONFIG_TICKLESS_KERNEL
+	u64_t next = last_count + CYC_PER_TICK;
+
+	if ((s64_t)(next - curr) < MIN_DELAY) {
+		next += CYC_PER_TICK;
```

Comment:
This breaks if MIN_DELAY is larger than one tick.  Not a problem given the numbers above, I think.  Might be worth a BUILD_ASSERT just to be sure.

Reply:
Added a build assert in the driver.